AF-2432 :  Restoring an old asset version doesn't work outside master branch

akumar074 opened the pull request:
* Added CurrentBranch class to track current modifying branch in the editor  

Related PR : 
kie-wb-common [#3238](https://github.com/kiegroup/kie-wb-common/pull/3238)

akumar074 commented:
Jenkins execute full downstream build please

akumar074 commented:
Jenkins execute full downstream build please

akumar074 commented:
Jenkins execute full downstream build please

akumar074 commented:
Jenkins execute full downstream build please

sthundat commented:
Jenkins execute full downstream build

akumar074 commented:
Jenkins retest this please

sthundat commented:
@akumar07 artifacts are not getting generated.Could you please take a look

akumar074 commented:
Jenkins execute full downstream build please

akumar074 commented:
Jenkins execute full downstream build please

akumar074 commented:
Jenkins execute full downstream build please

akumar074 commented:
@sthundat @pefernan , As FDB seems to be failing because of unrelated Test failure, I have added the link to JBPM-server-distribution zip in the description for you to test.

akumar074 commented:
Jenkins execute full downstream build please

akumar074 commented:
Jenkins execute full downstream build please

sthundat commented:
@akumar074 when i select the previous version of an asset from any branch(master/new),getting some server error.Please find the attachment  for details
![branchVersionIssue](https://user-images.githubusercontent.com/56147477/77903125-fc828480-729f-11ea-9369-9c397f9065fa.png)

akumar074 commented:
> @akumar074 when i select the previous version of an asset from any branch(master/new),getting some server error.Please find the attachment for details

@sthundat I followed your steps, but I am unable to reproduce this issue.

akumar074 commented:
Jenkins execute full downstream build please

akumar074 commented:
Jenkins execute full downstream build please

sthundat commented:
![Branchrestore](https://user-images.githubusercontent.com/56147477/77992203-3eb0d200-7343-11ea-807f-ea73ba9925a0.gif)
Please check the video

akumar074 commented:
> Please check the video
I followed these steps but I'm still unable to reproduce this.

ederign commented:
@caponetto can you please take a look at this?

sthundat commented on uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/common/CurrentBranchImpl.java:
```diff
@@ -0,0 +1,28 @@
+
+public class CurrentBranchImpl implements CurrentBranch {
+
+    private final String DEFAULT_BRANCH = "master";
```
please make this filed static

akumar074 replied:
Done

akumar074 commented:
Jenkins execute full downstream build please

akumar074 commented on uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-backend/src/main/java/org/uberfire/ext/editor/commons/backend/version/VersionServiceImpl.java:
```diff
@@ -70,12 +70,14 @@ public Path getPathToPreviousVersion(String uri) {
 
             final org.uberfire.java.nio.file.Path path = pathResolver.resolveMainFilePath(convert(_path));
-            final org.uberfire.java.nio.file.Path target = path.getFileSystem().getPath(path.toString());
+            final org.uberfire.java.nio.file.Path target = path.getFileSystem().getPath(currentBranch,
```
This is a good suggestion. I have changed the use of getPath from FS to VersionUtil. We can get the same target path now.

akumar074 commented on uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-backend/src/main/java/org/uberfire/ext/editor/commons/backend/version/VersionServiceImpl.java:
```diff
@@ -70,12 +70,14 @@ public Path getPathToPreviousVersion(String uri) {
     public Path restore(final Path _path,
-                        final String comment) {
+                        final String comment,
+                        final String currentBranch) {
```
I agree `branchName` would be more appropriate for other situations.

akumar074 commented on uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/BasicFileMenuBuilder.java:
```diff
@@ -77,7 +78,7 @@ BasicFileMenuBuilder addCopy(final PathProvider provider,
     BasicFileMenuBuilder addValidate(final Command command);
 
-    BasicFileMenuBuilder addRestoreVersion(final Path path);
+    BasicFileMenuBuilder addRestoreVersion(final Path path, final CurrentBranch currentBranch);
```
Done.

akumar074 commented on uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/BasicFileMenuBuilderImpl.java:
```diff
@@ -343,8 +344,8 @@ public BasicFileMenuBuilder addValidate(final Command validateCommand) {
     @Override
-    public BasicFileMenuBuilder addRestoreVersion(final Path path) {
-        this.restoreCommand = restoreVersionCommandProvider.getCommand(path);
+    public BasicFileMenuBuilder addRestoreVersion(final Path path, final CurrentBranch currentBranch) {
```
Done.

akumar074 commented on uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/common/CurrentBranchImpl.java:
```diff
@@ -0,0 +1,28 @@
+
+package org.uberfire.ext.editor.commons.client.menu.common;
+
+public class CurrentBranchImpl implements CurrentBranch {
```
Thanks @caponetto for this one. I was unaware of this annotation. It is really helpful.

akumar074 commented on uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/BaseEditor.java:
```diff
@@ -792,4 +793,8 @@ public VersionRecordManager getVersionRecordManager() {
+
+    public void setCurrentBranch(final CurrentBranch currentBranch) {
+        versionRecordManager.setCurrentBranch(currentBranch);
+    }
```
Removed from the VersionRecordManager class.

akumar074 commented on uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/history/VersionRecordManager.java:
```diff
@@ -296,6 +302,14 @@ private void onRestore(final @Observes RestoreEvent restore) {
+    public CurrentBranch getCurrentBranch() {
+        return currentBranch;
+    }
+
```
Removed from the VersionRecordManager class.

akumar074 commented on uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/history/VersionRecordManager.java:
```diff
@@ -226,10 +230,12 @@ private VersionRecord getCurrentVersionRecord() {
             restorePopUpPresenter.show(getCurrentPath(),
-                                       getCurrentVersionRecordUri());
+                                       getCurrentVersionRecordUri(),
+                                       getCurrentBranch());
```
Done.

akumar074 commented on uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/history/VersionRecordManager.java:
```diff
@@ -226,10 +230,12 @@ private VersionRecord getCurrentVersionRecord() {
             restorePopUpPresenter.restoreCommand(getCurrentPath(),
-                                                 getCurrentVersionRecordUri())
+                                                 getCurrentVersionRecordUri(),
+                                                 getCurrentBranch())
```
Done.

akumar074 commented:
Jenkins execute full downstream build please

akumar074 commented:
> Hi @akumar074
> I've added some comments on the PRs.
> Please check if they make sense and let me know if you have any questions.
> 
> I see that most of the changes on the PRs is because there is no way of getting the branch name from the `_path` variable in `VersionService#restore` method. We only have the version, and this version could be part of one or more branches. I'm gonna think a little bit more about it but probably we can't avoid these changes.

@caponetto , I have incorporated changes according to your CR. Please find my comments on the same.

akumar074 commented:
Jenkins execute full downstream build please

sthundat commented:
Functionality is working fine hence approving this PR
